### PR TITLE
Riscv64linux Test fixes

### LINKED ISSUE #78911: Fix cause of link failures in CI that necessitates ignore-32bit in run-make tests
Some run-make tests require an `ignore-32bit` directive as a workaround for link failures that occur during CI. #76256 and #77901 added or modified tests to include this directive in order to avoid at least this error: `arm-none-eabi-gcc: error: unrecognized command line option '-m64'`.

I assume that this shouldn't be necessary, and that there is actually some deficiency in our CI build system that we can fix.

### PR DESCRIPTION
**Pull request by tblah**
Get tests passing again using the riscv64gc-unknown-linux-gnu docker image.

Test with
```
src/ci/docker/run.sh riscv64gc-linux
```

## linkcheck
Linkcheck tests that interdocument links in the documentation are correct. Some interdocument links go between rustc and tools (such as rustdoc and cargo). When cross compiling, rustc is built for the host while some tools are built for the target. This goes for the documentation too. Because of this, links in the rustc documentation reffering to cargo or rustdoc documentation look broken.

This issue is worked around by disabling linkcheck for cross compilation builds.

## run-make tests

#78911 seems to happen because `--target` was not passed to `rustc`, but the target linker was specified, causing the target linker to be called with options intended for the host.

Resolves #78911 

In a separate issue, `issue-36710` was trying to run a binary built for the target on the host system. This will not work for any platform using `remote-test-server`/`client` (such as riscv64). I don't know of a way of skipping those platforms specifically, so I set this test to skip only on riscv64 for now.

### CODE REVIEW

File: src/test/run-make/incr-prev-body-beyond-eof/Makefile [hunk 1]
```diff
@@ -2,6 +2,7 @@ include ../../run-make-fulldeps/tools.mk
 
 # FIXME https://github.com/rust-lang/rust/issues/78911
 # ignore-32bit wrong/no cross compiler and sometimes we pass wrong gcc args (-m64)
+# ignore-riscv64 unrecognized command line option '-m64'
```
Comment (nagisa) [inline src/test/run-make/incr-prev-body-beyond-eof/Makefile:5]: Would be good to know why this argument gets to the linker in the first place – the riscv64 target does not appear to specify such an argument the way x86_64 targets do, for instance.

File: src/bootstrap/test.rs [hunk 2]
```diff
@@ -122,7 +122,17 @@ impl Step for Linkcheck {
...
+
+        // if we have different hosts and targets, some things may be built for
+        // the host (e.g. rustc) and others for the target (e.g. std). The
+        // documentation built for each will contain broken links to
+        // docs built for the other platform (e.g. rustc linking to cargo)
+        if builder.hosts == builder.targets {
+            run.default_condition(builder.config.docs)
+        } else {
+            run.default_condition(false)
+        }
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs:135]: I don't quite understand what you're saying here - I would expect our docs to not be missing parts when generated for a given host or target? Can you elaborate more on exactly what you're seeing (e.g., what broken link)?

File: src/test/run-make/incr-prev-body-beyond-eof/Makefile (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/test/run-make/incr-prev-body-beyond-eof/Makefile:5]: Yeah, I'd prefer to not patch over this - we should figure out why this is getting through. It looks like it's plausibly also a problem on 32-bit targets for some reason.

File: src/bootstrap/test.rs (hunk 2 above)
Comment (tblah) [reply src/bootstrap/test.rs:135]: rustc docs are built into `build/x86_64-unknown-linux-gnu/doc`. All other docs are built to `build/riscv64gc-unknown-linux-gnu/doc/`, so when linkcheck looks at the rustc interdoc links it doesn't find the other documentation.

The missing links are 
```
rustc/lints/listing/warn-by-default.html:387: broken link - rustdoc/lints.html
rustc/lints/listing/warn-by-default.html:1037: broken link - rustdoc/lints.html
rustc/lints/listing/warn-by-default.html:1250: broken link - rustdoc/lints.html
rustc/lints/listing/warn-by-default.html:1518: broken link - rustdoc/lints.html
rustc/lints/listing/allowed-by-default.html:416: broken link - rustdoc/lints.html
rustc/lints/listing/allowed-by-default.html:568: broken link - rustdoc/lints.html
rustc/lints/listing/allowed-by-default.html:603: broken link - rustdoc/lints.html
rustc/lints/listing/allowed-by-default.html:709: broken link - rustdoc/lints.html
rustc/index.html:171: broken link - cargo/index.html
rustc/what-is-rustc.html:171: broken link - cargo/index.html
rustc/print.html:173: broken link - cargo/index.html
rustc/print.html:221: broken link - reference/conditional-compilation.html
rustc/print.html:250: broken link - reference/items/external-blocks.html
rustc/print.html:273: broken link - reference/linkage.html
rustc/print.html:276: broken link - reference/linkage.html
rustc/print.html:283: broken link - edition-guide/introduction.html
rustc/print.html:327: broken link - reference/conditional-compilation.html
rustc/print.html:414: broken link - reference/items/extern-crates.html
rustc/print.html:1051: broken link - rustdoc/lints.html
rustc/print.html:1203: broken link - rustdoc/lints.html
rustc/print.html:1238: broken link - rustdoc/lints.html
rustc/print.html:1344: broken link - rustdoc/lints.html
rustc/print.html:2077: broken link - rustdoc/lints.html
rustc/print.html:2727: broken link - rustdoc/lints.html
rustc/print.html:2940: broken link - rustdoc/lints.html
rustc/print.html:3208: broken link - rustdoc/lints.html
rustc/print.html:4926: broken link - reference/conditional-compilation.html
rustc/print.html:5176: broken link - reference/expressions/operator-expr.html
rustc/print.html:5306: broken link - reference/attributes/codegen.html
rustc/print.html:5310: broken link - reference/linkage.html
rustc/codegen-options/index.html:210: broken link - reference/conditional-compilation.html
rustc/codegen-options/index.html:460: broken link - reference/expressions/operator-expr.html
rustc/codegen-options/index.html:590: broken link - reference/attributes/codegen.html
rustc/codegen-options/index.html:594: broken link - reference/linkage.html
rustc/command-line-arguments.html:173: broken link - reference/conditional-compilation.html
rustc/command-line-arguments.html:202: broken link - reference/items/external-blocks.html
rustc/command-line-arguments.html:225: broken link - reference/linkage.html
rustc/command-line-arguments.html:228: broken link - reference/linkage.html
rustc/command-line-arguments.html:235: broken link - edition-guide/introduction.html
rustc/command-line-arguments.html:279: broken link - reference/conditional-compilation.html
rustc/command-line-arguments.html:366: broken link - reference/items/extern-crates.html
error-index.html:7: broken link - light1.51.0.css
error-index.html:8: broken link - rust.css
```

File: src/test/run-make/incr-prev-body-beyond-eof/Makefile (hunk 1 above)
Comment (tblah) [reply src/test/run-make/incr-prev-body-beyond-eof/Makefile]: It turns out this was the same issue as for 32-bit platforms. The problem was that we didn't call rustc with `--target` but did specify the linker for the target platform.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@tblah - can you please post your status on this PR?

**Comment by tblah** [comment]
@JohnCSimon ready for review

**Comment by tblah** [comment]
@Mark-Simulacrum @nagisa ping

**Comment by Mark-Simulacrum** [comment]
For linkcheck, I'm sort of still uncertain - I would expect us to build docs for both rustc and cargo (for example) for a given platform, not just one of them, and as such for links to work as expected. Does one of them perhaps need a host-only annotation?

**Comment by Mark-Simulacrum** [comment]
Ok, so after looking into the code a bit, it sounds like the problem is that rustc and rustdoc internal documentation (i.e., rustdoc run on those crates) is built for host platforms only, but other documentation (e.g., cargo, reference, etc. books) are built for targets too.

I don't think disabling linkcheck is the right fix - you have genuinely broken links in the generated docs. Rather, it seems like the problem is that building the default documentation (i.e., `x.py doc`) isn't generating working documentation with host platforms not included in the targets. I'm not sure what exactly to do about this -- the user's intent with `x.py doc --host x86_64-unknown-linux-gnu --target riscv64gc-unknown-linux-gnu` is not precisely clear, but currently does build sort of "what is asked for":

* a set of docs for the host parts
* a set of docs for target parts

This would quite clearly lead to link failures from the host to the target.

I can see several strategies for fixing this:

1. disable link checking in such configurations (this PR)
2. build target docs for the host parts if we're linking to them (i.e., add ensures(...) for them) manually
3. change link checking to run on targets, rather than hosts

I'm personally not even sure that we should do any of these 3, but of them, I feel like option 2 is the most appealing from a "building working documentation" policy -- it is also unfortunately going to be very error prone over time if we're not testing this kind of configuration as links get added. Option 1 is definitely the easiest -- I'd be ok landing it with a panic instead of silently passing, and recommending that folks that explicitly are OK with broken documentation pass --exclude src/tools/linkchecker to their x.py test invocation. Do you think that's reasonable?

**Comment by tblah** [comment]
@Mark-Simulacrum thanks for taking a look.

The `x.py` invocation is for the [riscv tests dockerfile](https://github.com/rust-lang/rust/blob/master/src/ci/docker/host-x86_64/disabled/riscv64gc-linux/Dockerfile#L101):
```sh
python3 ../x.py --stage 2 test --target riscv64gc-unknown-linux-gnu
```
Are you saying the intent of that is unclear? What should it be? It looks like [armhf-gnu](https://github.com/rust-lang/rust/blob/master/src/ci/docker/host-x86_64/armhf-gnu/Dockerfile#L82) uses `--host='' --target arm-unknown-linux-gnueabihf`. I'll try this to see if it fixes this specific usecase.

Thanks for the possible solutions to the general problem. I'll see what I can get working this weekend.

**Comment by Mark-Simulacrum** [comment]
Yeah, currently the test invocation is testing sort of an unusual combination- the host is specified implicitly as the build triple, but the target list doesn't include the host. That's at least plausibly not intended. It's possible just specifying host to be empty (i.e. only check the target bits, not the compiler for example) is what's actually intended.

**Comment by tblah** [comment]
I've updated the MR adding `--host=''` to the riscv dockerfile and following strategy 1 with the `panic!`.

Thanks for your help @Mark-Simulacrum

**Comment by tblah** [comment]
The most recent failure was because the x64_64-gnu-llvm-9 Dockerfile also left the host to implicitly be different to the target (in one instance, other invocations do specify `--host=''`). I am not sure if that was intentional or a bug. 9b23df1 fixes it. We hadn't seen the linkcheck issue from this invocation because it is limited to `ui` tests.

I've checked the other Dockerfiles for the same issue and they look okay.

**Comment by tblah** [comment]
@camelid thanks

@rustbot label: +S-waiting-on-review -S-waiting-on-author

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/83134#issuecomment-799031767
@bors r-

**Comment by tblah** [comment]
I think the problem was that the tests I re-enabled were never meant to work with `no-std` targets. I've set them to be skipped in this case db14627

@rustbot label: +S-waiting-on-review -S-waiting-on-author

**Comment by tblah** [comment]
Context: 2f7672f passes `--target` to `rustc` where it wasn't previously specified in some run-make tests. This has broken these tests on some architectures where the tests should never have passed in the first place.

In 053bd36 I have skipped the changed run-make tests on some more targets. These two (wasm and nvtptx64-nvidia-cuda) look like they might be made to work given the attention of somebody who knows how to set up the docker container appropriately and what the correct rustc invocation is. There is more information in the commit message.

@rustbot label: +S-waiting-on-review -S-waiting-on-author

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/83222#issuecomment-800811905
@bors r-

**Comment by tblah** [comment]
Another architecture ignore added. It doesn't look like there is a musl C++ compiler built by the docker container: `/musl-i586/bin/` contains only `musl-gcc` and no `musl-g++`.

Thanks for your patience everyone.

**Comment by tblah** [comment]
In the new version I have rebased on top of the new changes and solved the test failure for armhf-gnu.

**Comment by jethrogb** [comment]
See #83661

**Pull request merged**